content/named-tuples.md

```diff
@@ -0,0 +1,282 @@
+name for two different elements.
+
+Fields of named tuples can be selected by their name, as in the line `p.age < 18` above.
+
+Example:
+
+~~~ scala
+// This is an @main method
+@main def foo(x: Int): Unit =
+  println(x)
+~~~
+
+### Conformance
+
+The order of names in a named tuple matters. For instance, the type `Person` above and the type `(age: Int, name: String)` would be different, incompatible types.
+
+Values of named tuple types can also be be defined using regular tuples. For instance:
+```scala
+val x: Person = ("Laura", 25)
+
+def register(person: Person) = ...
+register(person = ("Silvain", 16))
+register(("Silvain", 16))
+```
+This follows since a regular tuple `(T_1, ..., T_n)` is treated as a subtype of a named tuple `(N_1 = T_1, ..., N_n = T_n)` with the same element types. On the other hand, named tuples do not conform to unnamed tuples, so the following is an error:
+```scala
+val x: (String, Int) = Bob           // error: type mismatch
+```
+One can convert a named tuple to an unnamed tuple with the `toTuple` method, so the following works:
+```scala
+val x: (String, Int) = Bob.toTuple // ok
+```
+
+_Question:_ Should we define an implicit conversion, either in place of this method or in addition to it?
+
+Note that conformance rules for named tuples are analogous to the rules for named parameters. One can assign parameters by position to a named parameter list.
+```scala
+  def f(param: Int) = ...
+  f(param = 1)   // OK
+  f(2)           // Also OK
+```
+But one cannot use a name to pass an argument to an unnamed parameter:
+```scala
+    val f: Int => T
+    f(2)         // OK
+    f(param = 2) // Not OK
+```
+The rules for tuples are analogous. Unnamed tuples conform to named tuple types, but the opposite does not hold.
+
+
+### Pattern Matching
+
+When pattern matching on a named tuple, the pattern may be named or unnamed.
+If the pattern is named it needs to mention only a subset of the tuple names, and these names can come in any order. So the following are all OK:
+```scala
+Bob match
+  case (name, age) => ...
+
+Bob match
+  case (name = x, age = y) => ...
+
+Bob match
+  case (age = x) => ...
+
+Bob match
+  case (age = x, name = y) => ...
+```
+
+### Expansion
+
+Named tuples are in essence just a convenient syntax for regular tuples. In the internal representation, a named tuple type is represented at compile time as a pair of two tuples. One tuple contains the names as literal constant string types, the other contains the element types. The runtime representation of a named tuples consists of just the element values, whereas the names are forgotten. This is achieved  by declaring `NamedTuple`
+in package `scala` as an opaque type as follows:
+```scala
+  opaque type NamedTuple[N <: Tuple, +V <: Tuple] >: V = V
```

Comment:
Is there any way we constrain the `N` type parameter more, given that it can only be literal strings, and not arbitrary types like the `V` paramater next to it?

Reply:
The compiler checks that all labels are string literals. But we can't encode such a constraint in the type bound.

Reply:
I think the question is since we are relying on opaque types to represent `NamedTuples`, what stops the user from creating a badly formed NamedTuples manually or via macros? The Scala 3 compiler protects us from creating bad trees, but it won't in this case, unless the checker special-cases `NamedTuples` opaque types in its checks.

Reply:
the consequence would be like passing around `Tuple` instead of a concrete case

Reply:
It's probably worth listing out in the proposal all the "magic" constraints and checks that we are adding here, since they are not part of the normal type system and would be special-case support for `scala.NamedTuple` . From what I understand:

- `N <: Tuple` must only contain literal strings
- `N <: Tuple` and `V <: Tuple` must have the same length
- `N <: Tuple` cannot contain duplicates

Is that all, or are there other constraints I'm missing?

Reply:
That should be it.

Reply:
In fact, those constraints are not enforced by the compiler. You can form arbitrary `NamedTuple` types yourself. There's no additional magic. It's just that they won't fit any named tuple value you write if they are ill-formed.

Reply:
I think this would be a good occasion to add a `TupleOf[T]` (with the semantics `TupleOf[T]<: EmptyTuple | T *: TupleOf[T]`) to the standard library

While not necessary, it would make the meaning of `N` clearer (even if it is renamed to something else)